chalk-solve/src/clauses.rs

```diff
@@ -244,16 +244,17 @@ fn program_clauses_that_could_match<I: Interner>(
                 .opaque_ty_data(opaque_ty.opaque_ty_id)
                 .to_program_clauses(builder),
         },
-        DomainGoal::WellFormed(WellFormed::Trait(trait_predicate)) => {
-            db.trait_datum(trait_predicate.trait_id)
+        DomainGoal::WellFormed(WellFormed::Trait(trait_ref))
+        | DomainGoal::LocalImplAllowed(trait_ref) => {
+            db.trait_datum(trait_ref.trait_id)
                 .to_program_clauses(builder);
         }
+        DomainGoal::ObjectSafe(trait_id) => db.trait_datum(*trait_id).to_program_clauses(builder),
```

Comment:
According to the function documentation, we should be more precise in the clauses we generate in order to make solving more efficient. However, most of the branches go through general clause-generating functions which generate all possible clauses. Should this ObjectSafe goal follow along and go through the general function (current implementation) or simply check the database to see if it's object safe and add the fact if so?

Reply:
So, we don't *really* have to be all that precise with the number of clauses we generate, since all of them go through filtering before getting turned into `Strand`s. (See `program_clauses_for_goal`). However, here we probably *do* only want to generate just the fact.